web/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql

```diff
@@ -0,0 +1,426 @@
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+
+/*
+  MK_TRANSACTION
+  Side effects procedure.
+  Input: a concatenation of one or more DML sql statements split by semicolon.
+  It is important that the statements are not DDL.
+  Either all the statements in the block succeed and get committed or all get rolled back.
+  To drop:
+  DROP PROCEDURE {{.output_schema}}.mk_transaction(VARCHAR);
+*/
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.mk_transaction(DML_STATEMENTS VARCHAR)
+  RETURNS VARCHAR
+  LANGUAGE JAVASCRIPT
+  AS
+  $$
+
+  dmls = DML_STATEMENTS.split(';')
+                       .filter(function(stmt) {return stmt.trim() !== '';})
+                       .map(function(stmt) {return stmt.trim() + ';';});
+
+  snowflake.createStatement({sqlText: `BEGIN;`}).execute();
+  try {
+
+      dmls.forEach(function(stmt) {
+          snowflake.createStatement({sqlText: stmt}).execute();
+      });
+      snowflake.createStatement({sqlText: `COMMIT;`}).execute();
+
+  } catch(ERROR) {
+      snowflake.createStatement({sqlText: `ROLLBACK;`}).execute();
+
+      // Snowflake error is not very helpful here
+      var err_msg = "Transaction rolled back.Probably failed: " + DML_STATEMENTS + " :Error: ";
+      throw Error(err_msg + ERROR);
+  }
+
+  return "ok. Statements in transaction succeeded."
+
+  $$
+;
+
+
+/*
+  COLUMN_CHECK
+  Checks for mismatched columns between source and target tables.
+  If source table is missing columns, it errors.
+  If AUTOMIGRATE is 'TRUE',
+    it allows the target table to get added the additional columns,if any, in source.
+  Since ALTER TABLE is a DDL statement, it will be its own transaction.
+  This means that even if it was placed inside another transaction,
+    it would commit that, and then start another(implicit) for its own execution.
+  So, also it cannot be explicitly rolled back.
+  This is why column_check is not part of the commit_with_metadata procedure.
+
+  Input:
+  SRC_SCHEMA:  the schema of the source table
+  SRC_TBL:     the source table name
+  TRG_SCHEMA:  the schema of the target table
+  TRG_TBL:     the target table name
+  AUTOMIGRATE: whether any extra columns in source table will be added in target table (only 'TRUE' enables)
+
+  To drop:
+  DROP PROCEDURE {{.output_schema}}.column_check(VARCHAR,VARCHAR,VARCHAR,VARCHAR,VARCHAR);
+*/
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.column_check(SRC_SCHEMA  VARCHAR,
```

Comment:
This is good stuff. I've found a slight wrinkle - it needs to make sure that new columns are in the right position (at the end of the table), and that existing columns all exist _in the right order_ - otherwise our insert will fail:

```
CREATE TABLE scratch.test1 AS(SELECT 1 AS one, 2 AS two);

CREATE TABLE scratch.test2 AS(SELECT 1 AS one, 3 AS three);

CREATE TABLE scratch.test3 AS(SELECT 1 AS one, 1.5 AS onepointfive, 2 AS two)

-- This will fail, as we expect it to:
CALL derived.column_check(
    UPPER('scratch'),
    UPPER('test2'),
    UPPER('scratch'),
    UPPER('test1'),
    'TRUE'
);

-- This will succeed, and will add 'onepointfive' column to the end of the table
CALL derived.column_check(
    UPPER('scratch'),
![screenshot 2021-02-01 at 18 11 51](https://user-images.githubusercontent.com/25744626/106499845-075a0700-64b9-11eb-9dc8-58683aa55b87.png)

    UPPER('test3'),
    UPPER('scratch'),
    UPPER('test1'),
    'TRUE'
);


SELECT * FROM scratch.test1
-- result attached
```

Reply:
Since `test3` is the source table, `test1` is the target, so any extra columns of `test3`, get added at the end of `test1`, with `test1`'s columns in place. Isn't it the expected output?

Reply:
I don't think so - because what happens when we then insert the data?

In this case, inserting `SELECT * FROM test3` into `test1` will fail, because it will attempt to put a decimal into an integer field. 

But even if the column types are compatible - you'll end up with the data in the wrong columns, without it being obvious why.

So I think it's best to enforce a rule that you can't add columns to the middle of an existing table, only the end. If you try to, the columnCheck should fail.

Reply:
I may be wrong or misunderstanding, but actually running this, creates the same column type for `onepointfive` in `test1` as it is in `test3`. So it seems that column types are compatible.
Also, adding columns, adds at the end(I haven't found an explicit guarantee on this on the Snowflake docs, but it seems to be the case). The `column_check` code doing this is:
```
function add_columns_to(sch, tbl, cols) {
      var alter_stmt = `ALTER TABLE ` + sch + `.` + tbl + ` ADD COLUMN ` + cols;
      snowflake.createStatement({sqlText: alter_stmt}).execute();

      return "ok. Columns added.";
  }
```
So, how could we enforce what `ALTER` does? 
Then, indeed a `select *` for insert will fail, but this is something that comes after the column check/migrate. And it would also fail even even if we could enforce the column order. That's why the `commit_with_metadata` uses an explicit select matching the target columns instead of a `select *`.

Reply:
Yes, ALTER TABLE ADD COLUMN will add to the end of the column list - you can't mutate an existing table to add columns in the middle. 

So the design pattern I suggest is that we only allow one to add a column to the end of your table, and if you try to do something different, then that should fail. If someone's doing this, I expect that they have either made a mistake (in which case it should fail), or they need the specific ordering of columns that they've provided (in which case they need to manage it manually). 

If we are reordering the columns for them, the behaviour is counterintuitive for both cases IMO. 

I think it also impairs readability of the code - if I'm tracing back from a production table that has been amended many times, I don't expect to see a different order of columns.

Those are the main considerations I have here. We can add it to the list to chat about.